### PR TITLE
add metadata refresh workflow

### PR DESCRIPTION
**Pull request by pnadolny13**
Note that this is not a scheduled job yet. I dont have a way to test it until some version of the workflow is in main. I'd like to merge this then iterate on it a bit in a second PR where I also turn the schedule on along with the extraction job https://github.com/meltano/hub/pull/1208.

- Downloads the raw json files extracted by https://github.com/meltano/hub/pull/1208. It defaults to only SDK variants right now.
- Merge the downloaded json files with the existing definitions
- Create branch, commit changes, open PR

The create-pull-request GH action requires some credentials to allow it to open a PR https://github.com/peter-evans/create-pull-request#action-inputs. Currently we're referencing `secrets.GH_PAT`.

> Note: If you want pull requests created by this action to trigger an on: push or on: pull_request workflow then you cannot use the default GITHUB_TOKEN. See the [documentation here](https://github.com/peter-evans/create-pull-request/blob/main/docs/concepts-guidelines.md#triggering-further-workflow-runs) for workarounds.

### CODE REVIEW

**Comment by pnadolny13** [comment]
@edgarrmondragon do you have experience with GH actions that open a PR? I think we'll need to assign some permissions to allow it. I know we have the Melty Bot system user also that might work for this use case, or even a separate Hub Bot system user just for this use case.

**Comment by pnadolny13** [comment]
@WillDaSilva do you have experience creating a PR from a workflow? I read that action docs again and it says:

> Note: If you want pull requests created by this action to trigger an on: push or on: pull_request workflow then you cannot use the default GITHUB_TOKEN. See the [documentation here](https://github.com/peter-evans/create-pull-request/blob/main/docs/concepts-guidelines.md#triggering-further-workflow-runs) for workarounds.

I think we will want the PR thats created to kick off the normal CI jobs. The docs suggest a personal access token is the only way to do that and link to https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token. What do you think about using the Melty Bot user you created or creating a new Hub Bot user?

**Comment by WillDaSilva** [comment]
> I think we will want the PR thats created to kick off the normal CI jobs. The docs suggest a personal access token is the only way to do that and then they link to https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token. What do you think about using the Melty Bot user you created or creating a new Hub Bot user?

@pnadolny13 We can use @MeltyBot for this using `$MELTYBOT_GITHUB_AUTH_TOKEN`. It's already available to this repository, and should be able to create PRs.

**Comment by pnadolny13** [comment]
@WillDaSilva awesome! Is that set as a repo secret so I can access it like `token: ${{ secrets.MELTYBOT_GITHUB_AUTH_TOKEN }}`?

**Comment by WillDaSilva** [comment]
> Is that set as a repo secret so I can access it like token: `${{ secrets.MELTYBOT_GITHUB_AUTH_TOKEN }}`?

@pnadolny13 Yes. Sorry for not providing it in that format in the first place.

File: .github/workflows/metadata-refresh.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  #   - cron: '0 0 * * *' # Run at midnight UTC
+
+jobs:
+
+  metadata_refresh:
+    name: Metadata Refresh
+    runs-on: ubuntu-latest
+    environment: ${{ github.event.inputs.environment || github.event_name == 'schedule' && 'production' || 'preview' }}
+    env:
+      AWS_S3_BUCKET: "${{secrets.HUB_METADATA_S3_BUCKET }}"
```
Comment (WillDaSilva) [inline .github/workflows/metadata-refresh.yml:24]:
```suggestion
      AWS_S3_BUCKET: "${{ secrets.HUB_METADATA_S3_BUCKET }}"
```

File: .github/workflows/metadata-refresh.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    permissions:
+      id-token: write # This is required for requesting the JWT
+    steps:
+    - uses: actions/checkout@v3.3.0
+      # TODO: will need this to trigger CI jobs
+      # with:
+      #   token: ${{ secrets.GH_PAT }}
```
Comment (WillDaSilva) [inline .github/workflows/metadata-refresh.yml:31]: I don't understand this comment. Is it stale? Why would we need a personal access token to checkout the repository, and what does that have to do with triggering CI jobs?

File: .github/workflows/metadata-refresh.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      uses: peter-evans/create-pull-request@v4
+      with:
+        token: ${{ secrets.MELTYBOT_GITHUB_AUTH_TOKEN }}
+        commit-message: Refresh plugin metadata ${{ steps.date.outputs.date }}
+        committer: GitHub <noreply@github.com>
+        author: ${{ github.actor }} <${{ github.actor }}@users.noreply.github.com>
+        signoff: false
+        branch: hub-bot-metadata-${{ steps.date.outputs.date }}-${{ github.run_id }}-${{ github.run_number }}-${{ github.run_attempt }}
+        delete-branch: true
+        title: '[Hub Bot] Refresh metadata ${{ steps.date.outputs.date }}'
```
Comment (WillDaSilva) [inline .github/workflows/metadata-refresh.yml:63]: To be consistent with our other repositories.

```suggestion
        title: 'chore: [Hub Bot] Refresh metadata ${{ steps.date.outputs.date }}'
```

**Review by WillDaSilva — COMMENTED** [review]
Has this been run yet? You mentioned wanting to get it into `main` to iterate on it. That's fine by me, but as an alternative you could create a fork of the repository, or a demo repository, then run this workflow there with your own personal access token instead of MeltyBot's.

File: .github/workflows/metadata-refresh.yml (hunk 2 above)
Comment (pnadolny13) [reply .github/workflows/metadata-refresh.yml:31]: Yep thats stale, I thought I had removed it.

**Comment by pnadolny13** [comment]
@WillDaSilva thanks for the review, I think I incorporated all your feedback.

> Has this been run yet? You mentioned wanting to get it into main to iterate on it. That's fine by me, but as an alternative you could create a fork of the repository, or a demo repository, then run this workflow there with your own personal access token instead of MeltyBot's.

It hasnt run yet but as far as I'm aware I cant actually trigger it until its in main. Once I merge it I can edit a branch and test executing it from that branch. I can try to do it from a personal repo if thats preferred but since we give S3 access to the workflow by assuming a role using `AWS_ROLE_TO_ASSUME` I thought it might be tricky. I'll give that a shot first and see if I'm able to test.

**Comment by pnadolny13** [comment]
@WillDaSilva update, I was able to test this [on my fork](https://github.com/pnadolny13/hub/pull/5) after swapping out some credentials. This should work assuming the credentials work as expected in this repo.

**Review by WillDaSilva — APPROVED** [review]

**Pull request merged**